python/baseline/services.py

```diff
@@ -273,21 +328,42 @@ def predict(self, tokens, **kwargs):
 
         """
         preproc = kwargs.get('preproc', 'client')
+        exporter_field_feature_map = kwargs.get('exporter_field_feature_map', {})  # if empty dict argument was passed
+        if not exporter_field_feature_map:
+            exporter_field_feature_map = {'tokens': 'text'}
         label_field = kwargs.get('label', 'label')
         tokens_seq, mxlen, mxwlen = self.batch_input(tokens)
         self.set_vectorizer_lens(mxlen, mxwlen)
+        # TODO: unlike classify, here we allow vectorizers even for preproc=server to get `word_lengths`.
+        # vectorizers should not be available when preproc=server.
         examples = self.vectorize(tokens_seq)
         if preproc == 'server':
-            examples['tokens'] = [" ".join([y['text'] for y in x]) for x in tokens_seq]
+            unfeaturized_examples = {}
+            for exporter_field in exporter_field_feature_map:
+                unfeaturized_examples[exporter_field] = [" ".join([y[exporter_field_feature_map[exporter_field]]
+                                                                   for y in x]) for x in tokens_seq]
+            unfeaturized_examples['word_lengths'] = examples['word_lengths']
+            examples = unfeaturized_examples
 
         outcomes = self.model.predict(examples)
+
         outputs = []
+        if hasattr(self.model, 'return_labels'):
+            return_labels = self.model.return_labels
+        else:
+            return_labels = False  # keeping the default tagger behavior
+
+        if not return_labels:
+            label_vocab = revlut(self.get_labels())
```

Comment:
This will involve iterating though the label space on every call to predict. It would be nice to cache results or move this to where it doesn't happen all the time.

The same applies to the other Services

Reply:
i think this was the default behavior in taggerservice predict and it just returns a previously stored dict. not sure if we would save a lot by changing the behavior.